wikipedia/track.py

```diff
@@ -113,7 +121,60 @@ def params(self):
             return self.params()
 
 
+class CreateQueryRulesetParamSource(ParamSource):
+    def __init__(self, track, params, **kwargs):
+        super().__init__(track, params, **kwargs)
+        self.query_ruleset_params = QueryRulesetParams(track, params)
+
+    def partition(self, partition_index, total_partitions):
+        return self
+
+    def params(self):
+        rules = []
+        for i in range(self.query_ruleset_params.ruleset_size):
+            # Note: `AccessibleComputing` is one of the _ids indexed in the dataset
+            rule = {
+                "rule_id": "rule_{{i}}",
+                "type": "pinned",
+                "criteria": [{"type": "exact", "metadata": "rule_key", "values": [random.choice(["match", "no-match"])]}],
+                "actions": {"ids": [random.choice(["AccessibleComputing", "pinned-miss"])]},
+            }
+            rules.append(rule)
+
+        return {"method": "PUT", "path": f"{QUERY_RULES_ENDPOINT}/{self.query_ruleset_params.ruleset_id}", "body": {"rules": rules}}
+
+
+class QueryRulesSearchParamSource(QueryIteratorParamSource):
+    def __init__(self, track, params, **kwargs):
+        super().__init__(track, params, **kwargs)
+        self.query_ruleset_params = QueryRulesetParams(track, params)
+
+    def params(self):
+        try:
+            query = next(self._queries_iterator)
+            # TODO Update this to use current syntax with 8.15.0+
+            return {
+                "method": "POST",
+                "path": "/_search",
+                "body": {
+                    "query": {
+                        "rule": {
```

Comment:
Note: This syntax will only work for versions 8.15.0+ - is there anything else we should do for these challenges?